Fix builds for macOS

eugene-babichenko opened the pull request:
This pull request introduces:

- Successful builds with macOS GCC (Homebrew GCC 7.4.0 & 9.1.0)
- Using `conan` to fetch dependencies instead of just using `git`
- Several modification in `CMakeLists.txt` for readability.

**WIP**

- macOS clang builds

carlocayos commented:
Please create a PR to a `develop` branch first

eugene-babichenko committed:
STRM2-16 GLOB_RECURSE for project source files

This makes CMakeLists.txt much more compact and readable.

eugene-babichenko committed:
STRM2-16 Use conan to download dependencies

Conan provides standardized approach to use manage dependencies and
using it is faster than downloading dependencies with git.

eugene-babichenko commented:
@carlocayos Done.

eugene-babichenko commented on src/infrastructure/utils/variadic_struct.h:
```diff
@@ -363,9 +364,9 @@ namespace nem2_sdk { namespace internal {
 	
 	template<size_t I> struct struct_field_by_index<I, NullStruct> {
-		static_assert(I == 0, "variadic struct field index is out of range");
+		static_assert(I < 0, "variadic struct field index is out of range");
```
Fixed after [this discussion](https://github.com/proximax-storage/cpp-xpx-chain-sdk/commit/d3ef2762cb67c9db3ce27f2b4ce975472d7c856b#diff-b17e2305ea701bc695ebdd1f49cb657bR366) with @pananton13

pananton reviewed (commented):
* Introducing conan is great. Please remove also ./cmake/CMakeLists-rapidjson.txt.in file as it is not used anymore.
* I think that it will be better to fix bug in json parser in a way I've mentioned in a [comment](https://github.com/proximax-storage/cpp-xpx-chain-sdk/pull/1#commitcomment-34153125) regarding this commit.
* Also I'm still against recursive globbing of sources for the reasons described in discussion. Suggest creating a SourceList.cmake file in ./cmake where variables holding sources are initialized and including it to CMakeLists.txt

eugene-babichenko committed:
STRM2-15 Fix builds with macOS GCC and clang

* On macOS it is required to explicitly enable 'safe' functions by
  turning on the __STDC_WANT_LIB_EXT1__ flag.
* `static_assert` must depend on the actual value provided to a
  template. If a condition can be computed at the compile time and is
  known to always fail, compilation will fail even if the invalid
  expression is never constructed.
* Some implicit conversions are not supported by macOS GCC and clang.

Tested with the following compilers:

* Built-in GCC & clang: Apple LLVM version 10.0.1 (clang-1001.0.46.4)
* Homebrew GCC 7.4.0
* Homebrew GCC 9.1.0

eugene-babichenko committed:
cmake: fix CMP0068 warning by using OLD policy

pananton commented on src/sdk/crypto/private_key.cpp:
```diff
@@ -19,7 +19,7 @@ namespace nem2_sdk {
 	PrivateKey::PrivateKey(const supplier<uint8_t>& generator)
 	{
-		std::generate_n(key_.begin(), key_.end(), generator);
+		std::generate_n(key_.begin(), (long int) key_.end(), generator);
```
It was a bug here - see ```generate_n``` signature here https://en.cppreference.com/w/cpp/algorithm/generate_n . Second parameter should be number of elements to generate, not the iterator to end. Your fix is incorrect too, please replace to ```key_.end() - key_.begin()```

eugene-babichenko commented:
@carlocayos Please don't merge it right now, I will also check the build on Windows 10 tomorrow.

pananton commented:
@eugene-babichenko I'm also stared to fix build on windows (needed to compile project from win pc for myself and found that it does not build actually). There are few errors there, I fix a few and willf ix remaining tommorow.

pananton commented:
@eugene-babichenko Btw, I will create issues for work that I plan to do myself in order to not intersect with your work.

eugene-babichenko commented:
@pananton13 I rebased this onto your recent changes. Please check if you are able to build it.

eugene-babichenko committed:
fix build: JSON parser and private key

eugene-babichenko committed:
Merge branch 'develop' into feature/clang-build
